### PR TITLE
pm: power_domain: enforce strict initialisation sequencing

### PR DESCRIPTION
**Pull request by JordanYates**
Enforce strict initialisation ordering on power domains that depend on each other. Domains higher in the "chain" MUST be initialised first. This is required so that dependent devices can correctly query the higher level device for their own power state.

Priority cannot be set in Kconfig as per usual because different instances of the same driver need different levels.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

### CODE REVIEW

File: dts/bindings/power-domain/power-domain-gpio.yaml [hunk 1]
```diff
@@ -30,3 +30,10 @@ properties:
     required: false
     default: 0
     description: Off delay time, in microseconds
+
+  init-priority:
+    type: int
+    default: 75
```
Comment (gmarull) [inline dts/bindings/power-domain/power-domain-gpio.yaml:36]: why 75? defaults need to be justified.

File: drivers/power_domain/power_domain_gpio.c [hunk 2]
```diff
@@ -109,17 +109,17 @@ static int pd_gpio_init(const struct device *dev)
...
+	static const struct pd_gpio_config pd_gpio_##id##_cfg = {	    \
+		.enable = GPIO_DT_SPEC_INST_GET(id, enable_gpios),	    \
+		.startup_delay_us = DT_INST_PROP(id, startup_delay_us),	    \
+		.off_on_delay_us = DT_INST_PROP(id, off_on_delay_us),	    \
+	};								    \
+	static struct pd_gpio_data pd_gpio_##id##_data;			    \
+	PM_DEVICE_DT_INST_DEFINE(id, pd_gpio_pm_action);		    \
+	DEVICE_DT_INST_DEFINE(id, pd_gpio_init, PM_DEVICE_DT_INST_GET(id),  \
+			      &pd_gpio_##id##_data, &pd_gpio_##id##_cfg,    \
+			      POST_KERNEL, DT_INST_PROP(id, init_priority), \
```
Comment (gmarull) [inline drivers/power_domain/power_domain_gpio.c:122]: this makes me think it's time to review and remove/re-think the init priorities, they're a pain and we have DT ords now... cc: @mbolivar-nordic

File: dts/bindings/power-domain/power-domain-gpio.yaml (hunk 1 above)
Comment (JordanYates) [reply dts/bindings/power-domain/power-domain-gpio.yaml:36]: Docs updated

File: drivers/power_domain/power_domain_gpio.c (hunk 2 above)
Comment (JordanYates) [reply drivers/power_domain/power_domain_gpio.c:122]: #22545 is the issue tracking that feature. I have done some work/thinking recently about how to achieve that.

Comment (carlescufi) [reply drivers/power_domain/power_domain_gpio.c:122]: @gmarull please take another look

**Comment by carlescufi** [comment]
@ceolin @gmarull ping

**Comment by carlescufi** [comment]
> @ceolin @gmarull ping

@ceolin @gmarull ping

File: dts/bindings/power-domain/power-domain-gpio.yaml [hunk 3]
```diff
@@ -30,3 +30,13 @@ properties:
     required: false
     default: 0
     description: Off delay time, in microseconds
+
+  init-priority:
```
Comment (gmarull) [inline dts/bindings/power-domain/power-domain-gpio.yaml:34]: for me it's a -1 to set init priorities in DT, this has nothing to do with hardware description. We should finally fix the init priority mess in Zephyr.

Comment (henrikbrixandersen) [reply dts/bindings/power-domain/power-domain-gpio.yaml:34]: I agree.

Comment (galak) [reply dts/bindings/power-domain/power-domain-gpio.yaml:34]: Agreed as well!

**Review by galak — CHANGES_REQUESTED** [review]
We should NOT introduce `init-priority` into DTS.  As mentioned, we should just fix the init order using the dependency graph we have in DTS.

**Comment by ceolin** [comment]
> Agreed as well! 

Yes, I think we have a consensus that here is not the right place to tackle the problem. Also, this would be very limited to this class of power domains.

**Review by ceolin — CHANGES_REQUESTED** [review]
I share others concerns.

**Comment by ceolin** [comment]
> We should NOT introduce `init-priority` into DTS. As mentioned, we should just fix the init order using the dependency graph we have in DTS.

We need things done with SYS_INIT() that are not available from DTS

**Comment by JordanYates** [comment]
Consensus is that this is not the way to fix this issue, even as a temporary hack.
Will work on progressing a solution from DTS directly.

**Comment by JordanYates** [comment]
First steps towards fixing this are in #49318

**Pull request closed without merging**